### ISSUE TITLE
thought: using tree-sitter for AST transformation/codegen

### ISSUE THREAD

**Issue by ikrima** [tree-sitter/tree-sitter#642]
# TLDR:
- idea: hijack grammar.js to specify AST transformation rules/codegen
- request: thoughts on possible problems/complications and easiest way to implement

# Problem:
I find myself writing a lot of boilerplate query/parser code to generate or extract data from a file.
But, most of the transformation can be expressed as grammar rules that do the codegen

For example, I have a dsl to describe kernel operators. 

**The dsl file describing the  operator**
```cpp
@Import(es2chops)
struct SmoothingChop_PrmBlk_t {
  @UI(Hidden)
  SmplRate_t chSmplRate;
  @UI(Min=0,Max=10)
  SmplIdx_t chStartSmplNum;
  SmplIdx_t chEndSmplNum;
  @UI(Toggle)
  int32_t guassOrAvg;
  @UI(Adv)
  SmplIdx_t filterWidth;
};

@OpExprFnMeta(name: "SmoothingChop", desc: "SmoothingChop: Smooth all the things")
void SmoothingChop_Opdef_eval(
  @Arg(In)  SmoothingChop_PrmBlk_t a_parms ,
  @Arg(In)  Channel_t              a_chIn  ,
  @Arg(Out) Channel_t              a_chOut );
```

**Generated code artifact:** the dsl is transformed into code to be consumed by the compiler along with various bindings
```cpp
struct SmoothingChop_opArgBlk {
  SmoothingChop_PrmBlk_t a_parms;
  Channel_t              a_chIn  ;
  Channel_t              a_chOut ;
};

void SmoothingChop_Opdef_eval_glue(void* a_argBlk) {
  SmoothingChop_opArgBlk* args = (SmoothingChop_opArgBlk*)(a_argBlk);
  SmoothingChop_Opdef_eval(args.a_parms, args._a_chIn, args.a_chOut);
}
```

**Generated json type decl:** but I also may need to generate a json of the types in the file that is consumed by the UI system
```json
{
  "type": "SmoothingChop_PrmBlk_t",
  "fields": [
    {"type": "SmplRate_t", "name": "chSmplRate"}      ,
    {"type": "SmplIdx_t",  "name": "chStartSmplNum"}  ,
    {"type": "SmplIdx_t",  "name": "chEndSmplNum"}    ,
    {"type": "int32_t",    "name": "guassOrAvg"}      ,
    {"type": "SmplIdx_t",  "name": "filterWidth"}     ,
  ],
}
```

# Discussion

My current approach is creating a bunch of queries and manual traversal of the parse tree but it's a chore and a lot of boilerplate. It'd be great if I could just specify the transformation rules in a grammar file/query and piggy back off all the existing code that tree-sitter has

Effectively, that would allow me to use treesitter as a codegen tool 😍

**Questions:**
- What's the easiest way to extend tree-sitter to do this? Easiest == shortest time to implement, abusive hacks ok!
- Other easier approaches. I've thought about only using tree-sitter to export S-Expressions and then writing the codegen in another language. Downside is adding another language + duplicating parsing the tree-sitter's grammar.js
- Unanticipated here be dragons with this approach

**Comment by maxbrunsfeld** [comment]
Ok, I think I understand better what you're trying to do. Right now, I can't think of any helpful pointers, in terms of re-using code from the Tree-sitter codebase.

> I forgot about the json type declarations. That actually seems really close to what i need and maybe i can combine that with the new extended query syntax

Yeah, this `node-types.json` is meant to help you to meta-program code that consumes the syntax tree. Maybe it can help?

Sorry I can't be of more assistance. I'm going to close this out for now, but feel free to open further issues if you encounter more specific problems while working on this. It sounds like a very cool project!

**Comment by ikrima** [comment]
for posterity in case someone stumbles on this, the "thing"/name for what I was reaching for is
a. a finally tagless interpreter
b. tying treesitter to it

Once I stumbled on it, everything became a lot easier. (Had right idea but was doing it against the grain trying to put this in the parser logic vs. grammar). The tree sitter part is pretty simple now: just walk down the tree and return the functional associated with the grammar rule bound to its children

Here are my notes lifted my journal; maybe it'll save someone some time & help point them in the right direction

---
# Tagless Shallow embedding approach to interpreters

## TLDR
The functional programming literature is obtuse/makes things harder than they seem (it's described in terms of final algebras vs object algebras, etc) but all it really is: 
> instead of using data types to describe language expressions, they use functions



## Why?
- extremely simple way of making an embedded DSL (don't have to write parser/lexer/etc)
- performance (statement evaluation is function eval and in static langs, those can get easily inlined)
- solution to type expression/extensible in both ways


## Notes

- **Key Idea**: Describe language grammar as functions that take a free-bound "evaluator" function param instead of data types (a "final algebra"/"object algebra")
	- ✔: Performance (no tag dispatch), allows for partial evaluation extension, and expression problem solution
- **Explanations**:
    - ELIU in C# https://higherlogics.blogspot.com/2008/09/mostly-tagless-interpreters-in-c.html
    - http://lambda-the-ultimate.org/node/4572
    - Papers: 
		http://okmij.org/ftp/tagless-final/JFP.pdf
		http://www.cs.utexas.edu/~wcook/Drafts/2012/ecoop2012.pdf
- **Sample implementations**:
    - Simple C++ example ![[tagless-final-interp-cpp]]
    - Implementations in different languages: https://i.cs.hku.hk/~bruno/oa/
    - C# (2008) 
        - Snippet: http://lambda-the-ultimate.org/node/4572#comment-72110
        - More complete: http://lambda-the-ultimate.org/node/2569#comment-43805
    	- Expanded Version (2009): https://higherlogics.blogspot.com/2009/06/mobile-code-in-c-via-finally-tagless.html
		- Advanced Query language in C#: https://higherlogics.blogspot.com/2019/09/building-query-dsl-in-c.html
	- Snippet with Pratt parser (http://lambda-the-ultimate.org/node/4572#comment-72110)

**Comment by gpetrov** [comment]
@ikrima from what I read you just need an AST transformer.
@maxbrunsfeld is that something within the scope of tree sitter - to transform from one AST grammar to another?

**Comment by maxbrunsfeld** [comment]
> is that something within the scope of tree sitter - to transform from one AST grammar to another?

Programmatically creating a new AST is out of scope of Tree-sitter - Tree-sitter is really about going the other direction - source code -> syntax tree.

**Comment by Ledenel** [comment]
> hey @maxbrunsfeld , you're spot on. I agree and don't think tree-sitter main should be extended to support this (and now in hindsight, the feature-request label is misleading).
> 
> **I guess my ideal is:**
> 
> * get some guidance on hackiest way to get proof of concept
> * play with it, validate this is actually useful
> * share results/depending on utility, can re-evaluate whether to go back and make a more full fledged component of my language workbench
> 
> **non-goals:**
> 
> * add it to tree-sitter main. 🙅‍♀️ to scope creep
> 
> **concretely:**
> 
> * Why not just write traversal code directly? Ans: I've been and it's a lot of boilerplate
> * Why not just write your own generic tree-sitter traverser? Ans: I started to but found myself reimplementing shittier versions of the tree-sitter grammar for production rules
> * Why not just modify the cli? Ans: Was my first attempt but having never touched Rust, it wasn't....productive 😅
> 
> so the crazytown idea is just have define another "grammar" or an extension and have treesitter output that.
> 
> **Current attempts**
> Abuse the 'fields' in the grammar and stuff codegen function pointer names. Then, my tree-sitter visitor just walks through these
> 
> ```js
> opexpr_tagspec: $ => seq(
>   field('codegen_OpExpr', '_OpExprFnMeta'),
>   '(',
>   seq('name:',field('codegenArg_0', $.str_lit)),
>   ',',
>   seq('desc:',field('codegenArg_1', $.str_lit)),
>   ')'),
> ```
> 
> ```c++
> vector<void(*)(array<string> const*) codegen_pfn_table; 
> treesitter_visitor(node, ) {
>   if (codegen_pfn_table.contains(node.name())) {
>     (codegen_pfn_table[node.name()])(node.getcodegenfields());
>   }
> }
> ```
> 
> Super hacky, works really well, but quickly running into actually needing the fields and brittle :)
> 
> 1. I forgot about the json type declarations. That actually seems really close to what i need and maybe i can combine that with the new extended query syntax
> 
> **The ask**
> Feedback on approach/if there's an easier way, and any pointers to specific internal treesitter code functions or codeblocks that would be helpful!

By doing some quick and dirty exploration, it seems that `src/grammer.json` contains almost anything you want to recover code from AST. From now on, I've checked tree-sitter-python, tree-sitter-php and tree-sitter-scala. 

Here are some insights:

## Easy task
- recover punctuations, keywords: just find whether the node type contains a `type=STRING`, insert `value` on demand.
- decide where should be whitespace/newline: find `type=BLANK` nodes.
- find where are customizable parts in AST: find `type=PATTERN` nodes and check the regex pattern.

I've not found whether there are more terminal types in grammer.json, just one of `[STRING, BLANK, PATTERN]`. 

the easy task indicates that we can strip out these nodes to make tree-sitter's concrete syntax tree more "abstract". If we store the content of customizable part we can get equalivant and valid code via this AST, so it is doable to transform and generate new code.

this project may also helpful. It generates Haskell type-safe AST from tree-sitter grammer definition.

https://github.com/github/semantic#technology-and-architecture


## Hard task (requires existing code corpus)
- pretty-print: there's obviously no information about pretty-printing in `grammer.json`, but you may build a statistic model from existing code corpus (for example, compute the frequency of white-space counts before and after an AST node type).
- generate valid customizable parts: it's hard to generate string from regex, but most regexes are quite simple. If you can convert the regex to a DFA(determined finite automata) then you can easily generate text by assigning a probabilty distribution $P(from state, to state)$ to state-transitions, which will turn this regex into an generative model. After that you can sample from the regex to produce texts. What's more, you can extend and fit this probability distribution on existing code corpus to generate more human-readable texts. You may check whether the regex is fully customizable by checking that if there's loops in state-transition graph. If no loops, it suggests that the regex actually represents a finite language, so it may not be what you want (like php's `FaLSe` literal). 
- co-reference of identifiers (decide whether identifiers should be different or same): no information but you can build a nlp co-reference model from existing code corpus by assuming all customizable parts with same node type and content in one file are co-references. The granularity of this co-reference resolving model can be file-level, section-level (all ast-nodes with `type=SYMBOL, minimal depth to root node, not starts with _`), block-level (ast-nodes contains a 'MULTIPLE' definition and childrens rarely contains '\n') and statement-level (ast-nodes rarely contains '\n').

## Dirty task (requires human in the loop)
- external scanners: they are written in C or CPP (`indent` in tree-sitter-python), both of those are turing-complete language, so you have to write boilterplates (inverse funcs) to generate right tokens, it is non-trivial to implement an exactly inverse func for an function.
